kernel/src/hil/uart.rs

```diff
@@ -46,6 +46,9 @@ pub enum Error {
 
     /// No error occurred and the command completed successfully
     CommandComplete,
+
+    /// Client requested transfer bigger than buffer
+    ClientRequestLargerThanBuffer,
```

Comment:
Hrm... This is a bummer.

**High bit**: I think we should keep this, with a TODO in the comment that we can get rid of it by changing the interface to `transmit`/`receive`.

So this is _only_ meaningful because we're passing a slice and length separately. _That's_ necessary, rather than the client just slicing the buffer, because the client must _release_ their ownership of the buffer, and with only a slice, there is no way for the client to recover the full buffer in the callback.

The cost is that we've relinquished the static guarantee that the logical length (encoding in a separate argument now) is valid.

In the past, we've discussed having a special slice type that holds both the original buffer *and* a slice subset such that the original buffer is always recoverable. I believe if we had this, it obviate the need for this error, right?

Such a type also has the benefit that the referred slice doesn't have to start at the beginning of the buffer...

E.g.:

```rust
struct RecoverableSlice { // bad name
    original: &'static [u8],
    base_idx: usize,
    length: usize,
}

impl  RecoverableSlice {
    fn new<R: RangeBounds>(buf: &'static mut [u8], range: R) -> Result<RecoverableSlice, &'static mut [u8]> {
        if check_range_for_buf(range) {
            Ok(RecoverableSlice { original: buf, base_idx: range.start(), length: buf.len })
        } else {
            Err(buf)
        }
    }

    pub fn base(&self) -> *const u8 { &self.buf[self.base_idx] as *const u8 }

    pub fn len(&self) -> usize { self.len }

    pub fn recover(self) -> &'static mut [u8] {
        self.original
    }
}
```

Reply:
I agree that having some generic buffer object would perhaps be the cleanest way to handle this case, as otherwise, we theoretically will have this new error code proliferate throughout any Host module that could be mis-used with these buf/len parameters.

I feel like I'm missing something about how your RecoverableSlice works, but one thought for a simple struct/interface would be like this:
```
struct Buffer {
	pub data: &'static [u8],
	len: usize
}

enum Error {
	LenExceedsBufferSize
}

impl Buffer {
    // used by main when initializing capsules with ownership
    pub fn new(buf: &'static mut [u8]) -> Buffer {
	Buffer { data: buf, len: 0 };
    }
    // used by whoever filled the buffer
    pub set_len(self, len: usize) -> Result<Buffer, Error>{
	if self.data.len() >= len {
		self.len = len;
		Ok(self)	
	}
	else {
		Err(Error::LenExceedsBufferSize)	
	}
    }	
}```
Clients on both ends still manipulate Buffer.data directly, but the set_len  interface keeps a buffer manipulator from putting in a silly number, and so Hosts libraries receiving these objects wouldn't have to verify silly inputs.